core/domain/story_services_test.py

```diff
@@ -368,6 +369,146 @@ def test_update_story_acquired_skill_ids(self):
         self.assertEqual(
             story.story_contents.nodes[0].acquired_skill_ids, ['skill_id'])
 
+    def test_exploration_context_model_is_modified_correctly(self):
```

Comment:
Try adding another test that's somewhat tricky, e.g. add node 1, add node 2, delete node 1, delete node 2, add node 1, delete node 1, add node 2 -- all in the same change list.

Also make sure that the lists compiled in apply_change_list() handle this case correctly (e.g. they don't end up with multiple copies of node 1 or 2).

Reply:
Done, added the test.
In the internal list, whenever an exp_id is removed (when node is deleted or exp is changed), I go through the added list and remove the occurrence of that exp in it, which this new test should test. 
Other than that, since I am not modifying the flow of actually adding or deleting a node from the story, we won't end up with duplicate nodes in a story.

Reply:
Ah, sorry, by "node" above I meant exploration ID. Sorry for the confusion! Does your comment still apply?

Reply:
Ah, in that case, the new test should test duplicates, since if duplicates were created and only one was deleted, the other would still exist, but this test tests against that by verifying nothing exists, once the node linked to exp is deleted, so it should be fine.

Reply:
But at each step of the operation, are the internal lists in apply_change_list() correct as well?

Reply:
Since the effect on the models are correct, those have to be correct as well, right? Also, as they are private variables, is there any way we can test them?

Reply:
Ah, no, sorry, I was unclear. I just wanted to ask you to check this manually, just to be sure. But I do not think we need to write additional backend tests to validate the private variables.

Reply:
Done, added a screenshot to a comment below with the array values after the changelist is applied.